### ISSUE TITLE
Tree shaking issue

### ISSUE THREAD

**Issue by johnparkTopQuad** [patternfly/patternfly-react#3155]
We noticed that even though we're only using a couple of icons and a single react component, we're getting ove 2MB of icons included in our production bundles. Is the tree shaking working as intended? Is there something we need to do configuration-wise on our end to take advantage of tree-shaking on the consuming side? We're using a custom webpack configuration that is similar in setup to the seed project in terms of how the loaders are managed.

**Comment by dgutride** [comment]
@johnparkTopQuad  - sorry I haven't been able to repro this locally - treeshaking is working when I test in webpack, it's possible this is a babel issue.  If you can post any additional info you get from your investigation, that would be helpful.

**Comment by riccardo-forina** [comment]
Syndesis suffer of the same problem. Or, in a smaller scale, the [example app in use-patternfly](https://github.com/riccardo-forina/use-patternfly/tree/master/example) shows the same behavior. Checked with [source-map-explorer](https://www.npmjs.com/package/source-map-explorer)

**Comment by johnparkTopQuad** [comment]
Hey I wanted to give an update so you don't think this ticket is dead. I've been poking around with Webpack and Patternfly and measuring bundle size differences and I've noted the following:

Patternfly Seed Project -
TSX branch :
    Before adding Wizard:
        All (768.66 KB)
        app.bundle.js (757.71 KB)
        support.bundle.js (10.94 KB)

    After adding Wizard:
        All (850.78 KB)
        app.bundle.js (839.84 KB)
        support.bundle.js (10.94 KB)

I'm not sure if this is the expected growth of importing a single Wizard component. 


My project that consumes Patternfly -
Current bundle size  -   7.49 MiB
Current bundle size without Patternfly -  4.06 MiB

Since we are really only importing a single Wizard component I think were getting a lot more than we need. I tried changing out the babel-loaders for ts loaders and replicating what I could from seed but my results were the same. 

At this point ill see if i can maybe get seed project tree shaking to break / keep messing with Webpack.

**Comment by seanforyou23** [comment]
Related to https://github.com/patternfly/patternfly-react/issues/3268

**Comment by evwilkin** [comment]
@johnparkTopQuad @riccardo-forina FYI - for react-icons you can use the `babel-plugin-transform-imports` plugin and `babel.config.js` setup laid out here:

https://github.com/patternfly/patternfly-react/pull/3448

Keeping this issue open as we work on react-core components, which will require a different solution due to the directory structure & export names.

**Comment by Hyperkid123** [comment]
@evwilkin @johnparkTopQuad @riccardo-forina  I was playing with the transform import plugin yesterday and I have a solution. Although it's not 100% perfect dues to some limitations based on the file dist structure and named exports. This will be a long one.

### TLDR 
I will post a sample of babel config which works but there will be additional work required to make this work in 100% cases
```js
require.extensions['.css'] = () => undefined;
const path = require('path')
const glob = require("glob")

const mapper = {
  TextVariants: 'Text'
}

module.exports = {
  "presets": [
    "@babel/preset-env",
    "@babel/preset-react"
  ],
  plugins: [
  [
    'transform-imports',
    {
      '@patternfly/react-core': {
        transform: (importName, matches) => {
          let res;
          const files = glob.sync(path.resolve(__dirname, `../../node_modules/@patternfly/react-core/dist/js/**/${mapper[importName] || importName}.js`))
          if(files.length > 0) {
            res = files[0]
          } else {
            throw `File with importName ${importName} does not exist`
          }
          res = res.replace(path.resolve(__dirname, '../../node_modules/'), '')
          res = res.replace(/^\//, '')
          return res;
        },
        preventFullImport: true,
        skipDefaultConversion: true,
      },
    },
    'react-core',
  ], [
    'transform-imports',
    {
      '@patternfly/react-icons': {
        transform: (importName, matches) => `@patternfly/react-icons/dist/js/icons/${importName.split(/(?=[A-Z])/).join('-').toLowerCase()}`,
        preventFullImport: true,
      },
    },
    'react-icons',
  ] ],
}
```

## Issue with multiple named exports in a single file
This is something that cannot be resolved just by some clever config. It is based on the PF file structure which is not perfect. Basically to have this work you would need 1 component/function/we you export in one file. There is no way around that.

Why? I will use the `Text` component as an example. Let's say I want to use it. But even with the transform plugin, my bundle will have `TextVariants` object in the build, because its in the same file as a `Text` component. For this particular component its that big of a deal, but I think there are around 44 files with more than one export. Usually, it has like 4-6 exports. This means any time you use any component from these files, you will also have the rest of them. The tree shaking will not help outside of esm env. I did test that.

### How to partially solve multiple named exports in one file?
Let's continue with the `Text` and `TextVariants` example. If I want to use `Text` its ok it will work with the transform plugin config. But `TextVariants` is a problem. There is no file called `TextVariants.js` It is a named export inside `Text.js`. For that, you will have to create a mapper, which will tell you in what file this named export is. @karelhala was testing some automated way how to create said mapper. Output could look something like this:
```js
// named exports mapper
const mapper = {
  TextVariants: 'Text'
}
...
// the transform function
const files = glob.sync(path.resolve(__dirname, `../../node_modules/@patternfly/react-core/dist/js/**/${mapper[importName] || importName}.js`))

```

This will point to the correct file.

**This will not remove all unused components/functions from your build**. It only shows a path to them.

As I said to achieve the most effective solution, PF would have to change its file structure and have **one export per file**. No way around that.

**Comment by Hyperkid123** [comment]
@johnparkTopQuad I think that you should read through the #3448 issue and the linked issue which has a lot of info why is this happening. Also, update to the latest version of PF4.

It is a fix only for icons, but in one of our apps on cloud services which I used for testing, the build size went from 5MB to 1.81MB. that is a considerable difference.

**Comment by evwilkin** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Tree shaking - move from this milestone & put on hold, possibly add Breaking Changes label](https://trello.com/c/oCPzk2TJ/50-tree-shaking-move-from-this-milestone-put-on-hold-possibly-add-breaking-changes-label)

**Comment by redallen** [comment]
Should be fixed by #3517 which got merged just now. If you see this issue in @patternfly/react-core@^3.135.0 please reopen.

**Comment by Hyperkid123** [comment]
@redallen @evwilkin @johnparkTopQuad 

I just wanted to show a result of these changes to react-icons and react-core. The build size of this one project I am testing all of this went from 5.8 MB to 2.8 MB parsed size.

Here you can see the build chunks:
![Screenshot from 2020-02-03 15-00-43](https://user-images.githubusercontent.com/22619452/73659524-b8577700-4696-11ea-83a3-08ce21e1c5de.png)

The tokens look a bit too much, but I don't know how that is built. I will check what is going on there later, but these changes helped immensely. Thank you for collaborating on this.

**Comment by johnparkTopQuad** [comment]
Hey just had a chance to circle back to this and update Patternfly/react-core and I noticed the Patternfly bundle size dropped from 5.12 to 2.32. This is a great improvement! Thanks so much for taking care of this!
